docs/logql.md

```diff
@@ -0,0 +1,159 @@
+# LogQL: Log Query Language
+
+Loki comes with its very own language for querying logs called *LogQL*. LogQL
+can be considered a distributed `grep` with labels for extra querying control.
+
+A basic LogQL query consists of two parts: the **log stream selector** and a
+**filter expression**. Because of how Loki is designed, a set of log streams
+must be selected using a [Prometheus](https://prometheus.io)-style log stream
+selector.
+
+The log stream selector will reduce the number of log streams to a manageable
+volume. Depending how many labels you use to filter down the log streams will
+affect the relative performance of the query's execution. The filter expression
+is then used to do a distributed `grep` over the retrieved log streams.
+
+### Log Stream Selector
+
+The log stream selector determines which log streams should be included in your
+query. The stream selector is comprised of one or more key-value pairs, where
+each key is a **log label** and the value is that label's value.
+
+The log stream selector is written by wrapping the key-value pairs in a
+pair of curly braces:
+
+```
+{app="mysql",name="mysql-backup"}
+```
+
+In this example, log streams that have a label of `app` and `name` with exact
+values of `mysql` and `mysql-backup` will be included in the query results.
+
+The `=` operator after the label name is a **label matching operator**. The
+following label matching operators are supported:
+
+- `=`: exactly equal.
+- `!=`: not equal.
+- `=~`: regex matches.
+- `!~`: regex does not match.
+
+Examples:
+
+- `{name=~"mysql.+"}`
+- `{name!~"mysql.+"}`
+
+The same rules that apply for [Prometheus Label
+Selectors](https://prometheus.io/docs/prometheus/latest/querying/basics/#instant-vector-selectors)
+apply for Loki log stream selectors.
+
+### Filter Expression
+
+After writing the log stream selector, the resulting set of logs can be filtered
+further with a search expression. The search expression can be just text or
+regex:
+
+- `{job="mysql"} |= "error"`
+- `{name="kafka"} |~ "tsdb-ops.*io:2003"`
+- `{instance=~"kafka-[23]",name="kafka"} != kafka.server:type=ReplicaManager`
+
+In the previous examples, `|=`, `|~`, and `!=` act as **filter operators** and
+the following filter operators are supported:
+
+- `|=`: Log line contains string.
+- `!=`: Log line does not contain string.
+- `|~`: Log line matches regular expression.
+- `!~`: Log line does not match regular expression.
+
+Filter operators can be chained and will sequentially filter down the
+expression - resulting log lines must satisfy _every_ filter:
+
+`{job="mysql"} |= "error" != "timeout"`
+
+When using `|~` and `!~`,
+[RE2 syntax](https://github.com/google/re2/wiki/Syntax) regex may be used. The
+matching is case-sensitive by default and can be switched to case-insensitive
+prefixing the regex with `(?i)`.
+
+### Future Query Language Extensions
```

Comment:
Is it worth to document what _may_ be added in the future? Given the assumption "it's not done until it's done", we may give false expectations to the reader.

Reply:
Fair point :slightly_smiling_face: